sycl/doc/design/PropagateCompilerFlagsToRuntime.md

```diff
@@ -0,0 +1,130 @@
+was not able to set a breakpoint inside device code.
+
+## Requirements
+
+In order to support module-level debuggability, the user will compile different
+module files with different levels of optimization. These optimization levels
+must be preserved and made use of during the backend compilation. The following
+is a key requirement for this feature.
+- If the user specifies `-Ox` as a front-end compile option for a particular
+module, this option must be converted to appropriate backend option and then
+propagated fo use during backend JIT compilation.
+
+The following table specifies the appropriate backend options for level-zero and
+OpenCL backends.
+
+| Front-end option | L0 backend option | OpenCL backend option |
+| ---------------- | ----------------- | --------------------- |
+|      -O0         |  -ze-opt-disable  |   -cl-opt-disable     |
+|      -O1         |  -ze-opt-level=1  |   /* no option */     |
+|      -O2         |  -ze-opt-level=1  |   /* no option */     |
+|      -O3         |  -ze-opt-level=2  |   /* no option */     |
+
+
+## Proposed design
+
+This chapter discusses changes required in various stages of the compilation
+pipeline.
+
+
+### Changes to the clang front-end
+
+For each function in SYCL device code, we add a new function attribute that is
+named `sycl-optlevel`. Value of this attribute is set to the optimization level
+which was used to compile the overlying module.
+
+### Changes to the sycl-post-link tool
+
+During device code split performed in the `sycl-post-link` tool, optimization
+level attribute `sycl-optlevel` is treated as an optional feature,
+i.e. device code split algorithm ensures that no kernels with different values
+of sycl-optlevel are bundled into the same device image. See also optional
+kernel features [design document](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md#changes-to-the-post-link-tool).
+The `sycl-post-link` tool also adds a new property into the 
+`SYCL/misc properties` property set for each device code module. This entry will
+be used to store the optimization level. Name of this property is `optLevel` and
+the value is stored as a 32-bit integer. If there is a module where the user did
+not specify an optimization module, there is no new entry in the property set.
+
+### Changes to the SYCL runtime
+
+In the SYCL runtime, the device image properties can be accessed to extract the
+associated optimization level. Once the optimization level is available, a query
+is made to the plugin to identify the correct backend option. This backend
+option is added to the existing list of compiler options and is sent to the
+backend.
+
+### Changes to the plugin
+
+A new plugin API has been added. It takes the frontend option string as input in
+string format and returns `pi_result`. The signature is as follows:
+
+```C++
+pi_result piPluginGetBackendOption(pi_platform platform,
+                                   const char *frontend_option,
```

Comment:
I think this PI API was generalized since I last reviewed the design, and it now takes `const char *frontend_option` instead of a numeric optimization level.

Although I understand the desire to generalize the PI API, it doesn't fit in with the rest of the design.  In particular, the `SYCL/misc properties` property `optLevel` is the numeric optimization level, not a string.  This makes it unclear how the SYCL runtime can obtain the `frontend_option`.  Does the SYCL runtime reconstruct the frontend option and assume it has the form "-O\<optlevel\>"?  I guess that could be OK, though it seems a little strange to have logic like this in the runtime.  If we ever decide to rename the option (which is admittedly unlikely), we would need to change the runtime code.

It seems like the design would be cleaner if we did one of two things:

1. Carry the generalization throughout, changing the `optLevel` property and the `sycl-optlevel` IR attribute into strings also.
2. Decide that we only really care about optimization level, and just use the numeric optimization level in all the APIs (including PI).

I'm not super opposed to the current design, though.  If we keep it this way, please update the SYCL runtime section to describe how the runtime recreates the frontend option.  Also, add some explanation to the PI plugin section explaining why it takes a string rather than a numeric optimization level.

Reply:
Hi @gmlueck 

This is a good finding. I agree that the generalization should be carried throughout. Let me make the change and post a patch.

Thanks

Reply:
Based on discussion with @gmlueck, we will keep the current design and address any generalization in future PRs.
Minor documentation changes have been made.

Thanks Greg..